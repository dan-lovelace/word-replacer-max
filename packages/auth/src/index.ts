import { logDebug } from "@worm/shared";

logDebug("auth loaded");
