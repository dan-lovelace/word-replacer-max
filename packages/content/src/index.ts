import {
  renderContent,
  runStorageMigrations,
  startContentListeners,
} from "./lib";

document.addEventListener("readystatechange", () => {
  renderContent("document state change");
});

runStorageMigrations();
startContentListeners();
