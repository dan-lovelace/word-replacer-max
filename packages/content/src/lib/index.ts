export * from "./listeners";
export * from "./migrations";
export * from "./render";
