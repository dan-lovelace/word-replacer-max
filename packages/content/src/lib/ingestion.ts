import { z } from "zod";

import { browser, sendConnectMessage } from "@worm/shared/src/browser";
import {
  ErrorableMessage,
  HTMLReplaceResponse,
  HTMLStringItem,
  WebAppMessage,
  WebAppMessageData,
  WebAppMessageKind,
} from "@worm/types/src/message";
import { ExportLink } from "@worm/types/src/popup";
import {
  ReplacementStyle,
  ReplacementStyleOption,
  ReplacementSuggest,
  RuleGroups,
} from "@worm/types/src/replace";
import {
  FocusRule,
  StorageArea,
  StorageVersion,
  SyncStorageNew,
  SyncStoragePreferences,
} from "@worm/types/src/storage";

interface IngestionEngineConfig {
  ignoreElements?: Set<string>;
  startNode?: HTMLElement;
}

interface CacheStatus {
  isProcessing: boolean;
  timestamp: number;
}

type CacheStorage = Awaited<ReturnType<StorageArea["get"]>>;

type IngestionStorage = Partial<SyncStorageNew>;

const ExportLinkSchema: z.ZodType<ExportLink> = z.object({
  identifier: z.number(),
  url: z.string(),
});

const FocusRuleSchema: z.ZodType<FocusRule> = z.object({
  field: z.union([z.literal("queries"), z.literal("replacement")]),
  matcher: z.string(),
});

const PreferencesSchema: z.ZodType<SyncStoragePreferences> = z.object({
  activeTab: z.union([
    z.literal("account"),
    z.literal("domains"),
    z.literal("options"),
    z.literal("rules"),
    z.literal("support"),
  ]),
  domainListEffect: z.union([z.literal("allow"), z.literal("deny")]),
  extensionEnabled: z.boolean(),
  focusRule: FocusRuleSchema,
});

const ReplacementStyleOptionSchema: z.ZodType<ReplacementStyleOption> = z.union(
  [
    z.literal("backgroundColor"),
    z.literal("bold"),
    z.literal("color"),
    z.literal("italic"),
    z.literal("strikethrough"),
    z.literal("underline"),
  ]
);

const ReplacementStyleSchema: z.ZodType<ReplacementStyle> = z.object({
  active: z.boolean(),
  backgroundColor: z.string(),
  color: z.string(),
  options: z.array(ReplacementStyleOptionSchema),
});

const ReplacementSuggestSchema: z.ZodType<ReplacementSuggest> = z.object({
  active: z.boolean(),
});

const RuleGroupsSchema: z.ZodType<RuleGroups> = z.object({
  active: z.boolean(),
});

const StorageVersionSchema: z.ZodType<StorageVersion> = z.union([
  z.literal("1.0.0"),
  z.literal("1.1.0"),
  z.literal("1.1.1"),
  z.literal("1.2.0"),
]);

const SyncStorageSchema: z.ZodType<SyncStorageNew> = z.object({
  domainList: z.array(z.string()),
  exportLinks: z.array(ExportLinkSchema),
  preferences: PreferencesSchema,
  replacementStyle: ReplacementStyleSchema,
  replacementSuggest: ReplacementSuggestSchema,
  ruleGroups: RuleGroupsSchema,
  storageVersion: StorageVersionSchema,
});

export interface IngestionEngineStorageConfig {
  area: StorageArea;
  ttlMs?: number;
}

class IngestionEngineStorage {
  private readonly ttlMs: number;

  private updatePromise: Promise<void> | null;
  private cache: IngestionStorage;
  private cacheStatus: CacheStatus;
  private area: StorageArea;

  constructor(config: IngestionEngineStorageConfig) {
    const { area, ttlMs = 20 } = config;

    this.cache = {};

    this.cacheStatus = {
      isProcessing: false,
      timestamp: 0,
    };

    this.area = area;
    this.ttlMs = ttlMs;
    this.updatePromise = null;
  }

  private async updateStorage(): Promise<void> {
    const currentStorage = await this.area.get();
    const parseResult = SyncStorageSchema.safeParse(currentStorage);

    if (parseResult.error) {
      throw new Error("Error parsing current storage");
    }

    this.cache = parseResult.data;
  }

  public async getStorage(): Promise<IngestionStorage> {
    const now = Date.now();

    if (this.cacheStatus.isProcessing || this.updatePromise) {
      await this.updatePromise;
      return this.cache;
    }

    if (now - this.cacheStatus.timestamp <= this.ttlMs) {
      return this.cache;
    }

    this.cacheStatus.isProcessing = true;
    this.updatePromise = this.updateStorage();

    await this.updatePromise;
    return this.cache;
  }
}

export class IngestionEngine {
  private readonly ignoreElements: Set<string>;
  private readonly startNode: HTMLElement;

  private batchContents: HTMLStringItem[] = [];
  private batchSize = 50;
  private isProcessingBatch = false;
  private isProcessingStorage = false;
  private store: IngestionStorage;
  private storageProvider: IngestionEngineStorage;

  private sentParents: Map<string, HTMLElement> = new Map();

  constructor(options: IngestionEngineConfig = {}) {
    const {
      // TODO: make some of these user-defined?
      // TODO: add `code` element?
      ignoreElements = ["input", "script", "style", "textarea"],
      startNode = document.documentElement,
    } = options;

    this.store = {};

    this.storageProvider = new IngestionEngineStorage({
      area: browser.storage.sync,
      ttlMs: 80,
    });

    this.ignoreElements = new Set(ignoreElements);
    this.startNode = startNode;

    this.init();
  }

  /**
   * Handles DOM mutations
   * @param mutations - Array of MutationRecord objects
   * @private
   */
  private async handleMutations(mutations: MutationRecord[]): Promise<void> {
    if (mutations.length === 0) return;

    await this.updateStorageFromCache();
    const processedNodes = new Set<HTMLElement>();

    for (const mutation of mutations) {
      if (mutation.type === "characterData") {
        this.processTextNode(mutation.target as Text, processedNodes);
      } else {
        this.processAddedNodes(Array.from(mutation.addedNodes), processedNodes);
      }
    }

    if (this.batchContents.length > 0) {
      await this.processBatch();
    }
  }

  /**
   * Initializes the replacer by performing initial replacements and setting up
   * observers
   * @private
   */
  private async init(): Promise<void> {
    // Get latest storage
    await this.updateStorageFromCache();

    if (!this.store.preferences?.extensionEnabled) {
      // Extension is disabled, do not continue
      return;
    }

    // Wait for the user's current page to become available
    await this.waitForDocument();

    // Start listeners
    this.listenForMessages();
    this.listenForMutations();

    // Replace entire document during initialization
    this.replaceAllNodes();
  }

  /**
   * Determines if a mutation is relevant for text replacement
   * @param mutation - The MutationRecord to check
   * @private
   */
  private isRelevantMutation(mutation: MutationRecord): boolean {
    if (mutation.type === "characterData") return true;

    return Array.from(mutation.addedNodes).some(
      (node) =>
        node.nodeType === Node.TEXT_NODE ||
        (node.nodeType === Node.ELEMENT_NODE && node.hasChildNodes())
    );
  }

  /**
   * Sets up a runtime message listener for communication within the extension
   * @private
   */
  private listenForMessages() {
    browser.runtime.onMessage.addListener((message) => {
      const event = message as WebAppMessageData<WebAppMessageKind>;

      if (event.targets !== undefined && !event.targets.includes("content")) {
        return;
      }

      switch (event.kind) {
        case "htmlReplaceResponse": {
          const { data, error } =
            event.details as ErrorableMessage<HTMLReplaceResponse>;

          // console.log("replace response data", data);

          // TODO: make sure large amounts of mutations do not block
          const updates: Promise<void>[] = [];

          if (data !== undefined) {
            // TODO: apply dom mutations from response

            for (const item of data) {
              const parent = this.sentParents.get(item.id);
              // console.log("got parent", parent);

              if (!parent) continue;

              const update = new Promise<void>((res) => {
                parent.style.outline = "1px solid red";
                // parent.dataset.noReplace = "true";

                this.sentParents.delete(item.id);
              });

              updates.push(update);
            }
          }

          Promise.all(updates);

          break;
        }
      }

      return undefined;
    });
  }

  /**
   * Sets up a MutationObserver to watch for DOM changes
   * @private
   */
  private listenForMutations(): void {
    const observer = new MutationObserver((mutations) => {
      const textMutations = mutations.filter((mutation) =>
        this.isRelevantMutation(mutation)
      );

      this.handleMutations(textMutations);
    });

    observer.observe(this.startNode, {
      characterData: true,
      childList: true,
      subtree: true,
    });
  }

  /**
   * Processes newly added nodes for text replacement
   * @param nodes - Array of nodes to process
   * @param processedNodes - Set of already processed nodes
   * @private
   */
  private async processAddedNodes(
    nodes: Node[],
    processedNodes: Set<HTMLElement>
  ): Promise<void> {
    for (const node of nodes) {
      if (node.nodeType === Node.TEXT_NODE) {
        // console.log("processing added", node.parentElement);
        this.processTextNode(node as Text, processedNodes);
      } else if (node.nodeType === Node.ELEMENT_NODE) {
        this.walkTextNodes(node as HTMLElement, processedNodes);
      }
    }

    // if (this.batchContents.length > 0) {
    //   await this.processBatch();
    // }
  }

  private async processBatch(): Promise<void> {
    if (this.isProcessingBatch || this.batchContents.length === 0) {
      return;
    }

    this.isProcessingBatch = true;

    try {
      const batchToProcess = [...this.batchContents];
      this.batchContents = [];

      sendConnectMessage(
        "content",
        "htmlReplaceRequest",
        {
          strings: batchToProcess,
          syncStorage: this.store,
        },
        ["background"]
      );
    } finally {
      this.isProcessingBatch = false;

      // Check if more items were added while processing
      if (this.batchContents.length >= this.batchSize) {
        await this.processBatch();
      }
    }
  }

  /**
   * Processes text nodes for replacement
   * @param textNode - The text node to process
   * @param processedParents - Set of already processed text node parents
   * @private
   */
  private async processTextNode(
    textNode: Text,
    processedParents: Set<HTMLElement>
  ): Promise<void> {
    if (!this.shouldProcessTextNode(textNode, processedParents)) {
      return;
    }

    const { parentElement } = textNode;
    processedParents.add(parentElement);

    const batchItem: HTMLStringItem = {
      html: parentElement.outerHTML,
      id: crypto.randomUUID(),
    };

    this.batchContents.push(batchItem);
    this.sentParents.set(batchItem.id, parentElement);

    // console.log(
    //   "processing",
    //   textNode.textContent,
    //   JSON.stringify(textNode.textContent),
    //   "in",
    //   parentElement
    // );

    if (this.batchContents.length >= this.batchSize) {
      await this.processBatch();
    }
  }

  /**
   * Checks if an element should be ignored for text replacement
   * @param element - The element to check
   * @private
   */
  private shouldIgnoreElement(element: HTMLElement): boolean {
    return (
      this.ignoreElements.has(element.tagName.toLowerCase()) ||
      element.getAttribute("data-no-replace") === "true"
    );
  }

  /**
   *
   * @param textNode - The text node to check
   * @param processedParents - Set of already processed text node parents
   * @returns
   */
  private shouldProcessTextNode(
    textNode: Text,
    processedParents: Set<HTMLElement>
  ): textNode is Text & { parentElement: HTMLElement } {
    const { parentElement } = textNode;
    const textContent = textNode.textContent || "";

    if (textContent.trim() === "") {
      return false;
    }

    if (
      !parentElement ||
      processedParents.has(parentElement) ||
      this.shouldIgnoreElement(parentElement)
    ) {
      return false;
    }

    return true;
  }

  /**
   * Updates the internal storage from cache
   * @private
   */
  private async updateStorageFromCache(): Promise<void> {
    if (this.isProcessingStorage) return;

    try {
      this.isProcessingStorage = true;
      this.store = await this.storageProvider.getStorage();
    } finally {
      this.isProcessingStorage = false;
    }
  }

  /**
   * Waits for the document to be ready
   */
  private async waitForDocument(): Promise<void> {
    await new Promise<void>((resolve, reject) => {
      const timeout = setTimeout(() => {
        reject(new Error("Document failed to load within timeout period"));
      }, 6000);

      if (document.readyState === "complete") {
        clearTimeout(timeout);
        resolve();
      } else {
        document.addEventListener("readystatechange", () => {
          if (document.readyState === "complete") {
            clearTimeout(timeout);
            resolve();
          }
        });
      }
    }).finally(() => {
      document.body.style.opacity = "1";
    });
  }

  /**
   * Recursively walks through text nodes in an element
   * @param element - The element to walk through
   * @param processedParents - Set of already processed text node parents
   * @private
   */
  private walkTextNodes(
    element: HTMLElement,
    processedParents: Set<HTMLElement>
  ): void {
    if (this.shouldIgnoreElement(element)) return;

    const walker = document.createTreeWalker(
      element,
      NodeFilter.SHOW_TEXT,
      null
    );

    let node: Text | null;
    while ((node = walker.nextNode() as Text)) {
      this.processTextNode(node, processedParents);
    }
  }

  /**
   * Performs an initial replacement on the entire start node
   * @public
   */
  public async replaceAllNodes(): Promise<void> {
    await this.updateStorageFromCache();

    const processedParents = new Set<HTMLElement>();
    this.walkTextNodes(this.startNode, processedParents);

    // process any remaining items in the queue
    if (this.batchContents.length > 0) {
      await this.processBatch();
    }
  }
}
