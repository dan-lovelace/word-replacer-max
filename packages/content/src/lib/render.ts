import { isDomainAllowed, replaceAll } from "@worm/shared";
import {
  getStylesheet,
  STYLE_ELEMENT_ID,
} from "@worm/shared/src/replace/lib/style";
import { storageGetByKeys } from "@worm/shared/src/storage";
import { Storage } from "@worm/types";

type Cacheable<T> = {
  expires: number;
  value?: T;
};

type RenderCache = {
  storage: Cacheable<Storage>;
  styleElement: Cacheable<HTMLStyleElement>;
};

const RENDER_STORAGE_CACHE_LENGTH_MS = 100;
const RENDER_STYLE_CACHE_LENGTH_MS = 1000;

const renderCache: RenderCache = {
  storage: {
    expires: 0,
  },
  styleElement: {
    expires: 0,
  },
};

export async function renderContent(msg = "") {
  const now = new Date().getTime();

  if (now > renderCache.storage.expires) {
    const storage = await storageGetByKeys([
      "domainList",
      "matchers",
      "preferences",
      "replacementStyle",
    ]);

    renderCache.storage.expires = now + RENDER_STORAGE_CACHE_LENGTH_MS;
    renderCache.storage.value = storage;

    if (now > renderCache.styleElement.expires) {
      const renderCacheExpires = now + RENDER_STYLE_CACHE_LENGTH_MS;
      const existingStyleElement = document.head.querySelector(
        `#${STYLE_ELEMENT_ID}`
      ) as HTMLStyleElement;

      if (existingStyleElement) {
        renderCache.styleElement = {
          expires: renderCacheExpires,
          value: existingStyleElement,
        };
      } else {
        const newStyleElement = getStylesheet(storage.replacementStyle);

        document.head.appendChild(newStyleElement);

        renderCache.styleElement = {
          expires: renderCacheExpires,
          value: newStyleElement,
        };
      }
    }
  }

  if (!renderCache.storage.value) {
    return;
  }

  const {
    storage: {
      value: { domainList = [], matchers = [], preferences, replacementStyle },
    },
  } = renderCache;

  if (preferences) {
    if (!preferences.extensionEnabled) {
      return;
    }

    if (!isDomainAllowed(domainList, preferences)) {
      return;
    }
  }

  replaceAll(matchers, replacementStyle);
}
