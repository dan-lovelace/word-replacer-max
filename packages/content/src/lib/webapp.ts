import {
  createWebAppMessage,
  getWebAppIFrame,
  logDebug,
  webAppMessages,
} from "@worm/shared";
import { getApiEndpoint } from "@worm/shared/src/api/endpoints";
import {
  WebAppMessage,
  WebAppMessageKind,
  WebAppMessageKindMap,
} from "@worm/types/src/message";
import { storageSetByKeys } from "@worm/shared/src/storage";
import { ApiAuthTokensResponse } from "@worm/types";

/**
 * Attempt to communicate with the running `webapp` package in whichever
 * environment the extension is deployed. Communication is done through an
 * iframe rendered in the webapp DOM using the `window.postMessage` API.
 *
 * Polling is used to look for the expected `iframe` element in the webapp DOM.
 * When found, a message is posted to its window by the content script letting
 * the webapp know it is ready for messages.
 */
export function initializeWebApp() {
  const initializationStarted = new Date().getTime();
  const initializationLengthMs = 10000;
  const initializationIntervalMs = 50;

  const initialization: ReturnType<typeof setInterval> = setInterval(() => {
    if (new Date().getTime() - initializationStarted > initializationLengthMs) {
      logDebug("Timeout initializing web app");
      return clearInterval(initialization);
    }

    const elementQuery = getWebAppIFrame();

    if (!elementQuery || !elementQuery.contentWindow) return;

    elementQuery.contentWindow.addEventListener(
      "message",
      async (event: WebAppMessage<WebAppMessageKind>) => {
        switch (event.data.kind) {
          case webAppMessages.AUTH_TOKENS: {
            const { data } = event.data.details as ApiAuthTokensResponse;
            const endpoint = getApiEndpoint("AUTH_WHOAMI");
            const response = await fetch(endpoint, {
              headers: {
                Authorization: `Bearer ${data?.accessToken}`,
                "Content-Type": "application/json",
              },
              method: "GET",
            });

            if (!response.ok) {
              sendWebAppMessage("showToastMessage", {
                message: "Authorization failed",
                options: {
                  severity: "danger",
                },
              });
            } else {
              sendWebAppMessage("showToastMessage", {
                message: "Login success",
                options: {
                  severity: "success",
                },
              });

              storageSetByKeys({
                authentication: {
                  tokens: event.data.details as ApiAuthTokensResponse,
                },
              });
            }

            break;
          }

          case webAppMessages.PING_REQUEST: {
            const latestElement = getWebAppIFrame();

            if (!latestElement || !latestElement.contentWindow) {
              logDebug(
                "Unable to locate iframe content window when responding to ping"
              );
              break;
            }

            sendWebAppMessage(webAppMessages.PING_RESPONSE, true);
            break;
          }

          default:
        }
      }
    );

    const newMessage = createWebAppMessage(
      webAppMessages.CONTENT_INITIALIZE,
      true
    );
    elementQuery.contentWindow.postMessage(newMessage);

    clearInterval(initialization);
  }, initializationIntervalMs);
}

export function sendWebAppMessage<T extends WebAppMessageKind>(
  kind: T,
  details?: WebAppMessageKindMap[T]
) {
  const latestElement = getWebAppIFrame();

  if (!latestElement || !latestElement.contentWindow) {
    return false;
  }

  const newMessage = createWebAppMessage(kind, details);
  latestElement.contentWindow.postMessage(newMessage);

  return true;
}
