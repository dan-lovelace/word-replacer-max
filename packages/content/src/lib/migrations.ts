import { merge } from "ts-deepmerge";

import {
  BASELINE_STORAGE_VERSION,
  CURRENT_STORAGE_VERSION,
  logDebug,
  STORAGE_MATCHER_PREFIX,
  storageGet,
  storageSet,
} from "@worm/shared";
import { DEFAULT_REPLACEMENT_STYLE } from "@worm/shared/src/replace/lib/style";
import { Matcher, Storage, StorageVersion, storageVersions } from "@worm/types";

type MigrateFn = (storage: Storage) => Storage;

async function fetchStorageVersion() {
  const { storageVersion } = await storageGet(["storageVersion"]);

  return storageVersion;
}

function isVersionGreaterThan(a: StorageVersion, b: StorageVersion): boolean {
  const [aMajor, aMinor, aPatch] = a.split(".").map(Number);
  const [bMajor, bMinor, bPatch] = b.split(".").map(Number);

  if (aMajor > bMajor) return true;
  if (aMajor < bMajor) return false;
  if (aMinor > bMinor) return true;
  if (aMinor < bMinor) return false;

  return aPatch > bPatch;
}

/**
 * Contains the transformation methods necessary to increment a single version.
 * If the current storage version is `1.0.0`, run `migrations[1][0][0] to apply
 * the next version.
 */
const migrations: {
  [major: number]: { [minor: number]: { [patch: number]: MigrateFn } };
} = {
  0: {
    0: {
      0: (storage) => {
        const updatedValues: Storage = {
          storageVersion: "1.0.0",
        };
        const merged = merge(storage, updatedValues);

        return merged;
      },
    },
  },
  1: {
    0: {
      0: (storage) => {
        const updatedValues: Storage = {
          replacementStyle: DEFAULT_REPLACEMENT_STYLE,
          storageVersion: "1.1.0",
        };
        const merged = merge(storage, updatedValues);

        /**
         * Add new properties to all stored matchers.
         */
        for (const key of Object.keys(merged)) {
          if (key.startsWith(STORAGE_MATCHER_PREFIX)) {
            const matcher = merged[key as keyof typeof merged] as Matcher;

            matcher.useGlobalReplacementStyle = true;
          }
        }

        return merged;
      },
    },
  },
};

/**
 * A hand-rolled database migration system for browser storage.
 */
export async function runStorageMigrations(
  againstVersion: StorageVersion = CURRENT_STORAGE_VERSION
) {
  let storageVersion = await fetchStorageVersion();

  if (!storageVersion) {
    /**
     * Storage does not contain a storage version at all. This must be a very
     * old installation so we need to start from scratch.
     */
    await storageSet({
      storageVersion: BASELINE_STORAGE_VERSION,
    });

    storageVersion = await fetchStorageVersion();
  }

  if (!storageVersion) {
    logDebug("Unable to obtain storage version");
    return;
  }

  if (!storageVersions.includes(storageVersion as StorageVersion)) {
    logDebug("Invalid storage version", storageVersion);
    return;
  }

  if (storageVersion === againstVersion) {
    /**
     * No migrations necessary.
     */
    return;
  }

  let migrationCount = 0;

  for (const version of storageVersions) {
    if (!storageVersion) continue;
    if (!isVersionGreaterThan(againstVersion, version)) continue;

    const [major, minor, patch] = version.split(".").map(Number);
    const migrateFn = migrations[major]?.[minor]?.[patch];

    if (!migrateFn) {
      continue;
    }

    const currentStorage = await storageGet();

    logDebug("Running migration", version);
    const newStorage = migrateFn(currentStorage);

    await storageSet(newStorage);

    migrationCount += 1;
    storageVersion = await fetchStorageVersion();
  }

  if (migrationCount > 0) {
    logDebug("Successfully applied", migrationCount, "migration(s)");
  }
}
