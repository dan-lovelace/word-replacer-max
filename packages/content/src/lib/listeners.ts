import { debounce, logDebug } from "@worm/shared";
import { browser, sendConnectMessage } from "@worm/shared/src/browser";
import { storageGetByKeys } from "@worm/shared/src/storage";
import {
  ErrorableMessage,
  HTMLReplaceResponse,
  WebAppMessageData,
  WebAppMessageKind,
} from "@worm/types/src/message";

import { renderContent } from "./render";

const render = debounce(renderContent, 20);

export function startContentListeners() {
  /**
   * Re-render whenever storage changes.
   */
  browser.storage.onChanged.addListener(() => renderContent());

  /**
   * Listen for changes to the document and render when they occur.
   */
  const observer = new MutationObserver((mutationList) => {
    const hasTextChanges = mutationList.some(
      (mutation) =>
        mutation.type === "characterData" ||
        Array.from(mutation.addedNodes).some(
          (node) => node.nodeType === Node.TEXT_NODE || node.hasChildNodes()
        )
    );

    if (hasTextChanges) {
      render("mutation");
    }
  });

  observer.observe(document.documentElement, {
    childList: true,
    subtree: true,
    characterData: true,
  });

  /**
   * Listen for runtime messages.
   */
  browser.runtime.onMessage.addListener((message) => {
    const event = message as WebAppMessageData<WebAppMessageKind>;

    if (event.targets !== undefined && !event.targets.includes("content")) {
      return;
    }

    switch (event.kind) {
      case "htmlReplaceResponse": {
        const { data, error } =
          event.details as ErrorableMessage<HTMLReplaceResponse>;

        console.log("replace response data", data);

        // TODO: apply dom mutations from response
        // TODO: make sure large amounts of mutations do not block

        break;
      }
    }

    return undefined;
  });

  storageGetByKeys()
    .then((syncStorage) => {
      // TODO: send initial strings for replacement

      sendConnectMessage(
        "content",
        "htmlReplaceRequest",
        {
          strings: [
            {
              html: "<div>hello</div>",
              id: crypto.randomUUID(),
            },
            {
              html: "<div>world</div>",
              id: crypto.randomUUID(),
            },
          ],
          syncStorage,
        },
        ["background"]
      );
    })
    .catch((error) => {
      logDebug("Failed to initialize storage in content script", error);
    });
}
