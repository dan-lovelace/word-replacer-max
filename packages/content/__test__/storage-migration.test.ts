import { expect } from "@jest/globals";

import { storageClear, storageGet, storageSet } from "@worm/shared/src/storage";

import { runStorageMigrations } from "../src/lib";

describe("runStorageMigrations", () => {
  beforeEach(async () => {
    await storageClear();
  });

  it("does not run migrations when on the latest version 1.0.0", async () => {
    const testStorage = {
      domainList: [],
      exportLinks: [],
      "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
        active: true,
        identifier: "b7fce47e-58e8-4409-adf4-08da053e713d",
        queries: ["my jaw dropped", "I was shocked"],
        queryPatterns: [],
        replacement: "I was surprised",
      },
      preferences: {
        activeTab: "rules",
        domainListEffect: "deny",
        extensionEnabled: true,
        focusRule: "",
      },
      storageVersion: "1.0.0",
    };

    await storageSet(testStorage);
    await runStorageMigrations("1.0.0");

    const result = await storageGet();
    expect(result).toMatchSnapshot();
  });

  it("does not run migrations when on the latest version 1.1.0", async () => {
    const testStorage = {
      domainList: [],
      exportLinks: [],
      "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
        active: true,
        identifier: "b7fce47e-58e8-4409-adf4-08da053e713d",
        queries: ["my jaw dropped", "I was shocked"],
        queryPatterns: [],
        replacement: "I was surprised",
        useGlobalReplacementStyle: true,
      },
      preferences: {
        activeTab: "rules",
        domainListEffect: "deny",
        extensionEnabled: true,
        focusRule: "",
      },
      replacementStyle: {
        active: false,
        backgroundColor: "#ffc107",
        color: "#dc3545",
        options: ["backgroundColor"],
      },
      storageVersion: "1.1.0",
    };

    await storageSet(testStorage);
    await runStorageMigrations("1.1.0");

    const result = await storageGet();
    expect(result).toMatchSnapshot();
  });

  it("migrates from undefined version to 1.0.0", async () => {
    const testStorage = {
      domainList: [],
      exportLinks: [],
      "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
        active: true,
        identifier: "b7fce47e-58e8-4409-adf4-08da053e713d",
        queries: ["my jaw dropped", "I was shocked"],
        queryPatterns: [],
        replacement: "I was surprised",
      },
      preferences: {
        activeTab: "rules",
        domainListEffect: "deny",
        extensionEnabled: true,
        focusRule: "",
      },
    };

    await storageSet(testStorage);
    await runStorageMigrations("1.0.0");

    const result = await storageGet();
    expect(result).toMatchSnapshot();
  });

  it("migrates from undefined version to 1.1.0", async () => {
    const testStorage = {
      domainList: [],
      exportLinks: [],
      "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
        active: true,
        identifier: "b7fce47e-58e8-4409-adf4-08da053e713d",
        queries: ["my jaw dropped", "I was shocked"],
        queryPatterns: [],
        replacement: "I was surprised",
      },
      preferences: {
        activeTab: "rules",
        domainListEffect: "deny",
        extensionEnabled: true,
        focusRule: "",
      },
    };

    await storageSet(testStorage);
    await runStorageMigrations("1.1.0");

    const result = await storageGet();
    expect(result).toMatchSnapshot();
  });

  it("migrates from 1.0.0 to 1.1.0", async () => {
    const testStorage = {
      domainList: [],
      exportLinks: [],
      "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
        active: true,
        identifier: "b7fce47e-58e8-4409-adf4-08da053e713d",
        queries: ["my jaw dropped", "I was shocked"],
        queryPatterns: [],
        replacement: "I was surprised",
      },
      preferences: {
        activeTab: "rules",
        domainListEffect: "deny",
        extensionEnabled: true,
        focusRule: "",
      },
      storageVersion: "1.0.0",
    };

    await storageSet(testStorage);
    await runStorageMigrations("1.1.0");

    const result = await storageGet();
    expect(result).toMatchSnapshot();
  });
});
