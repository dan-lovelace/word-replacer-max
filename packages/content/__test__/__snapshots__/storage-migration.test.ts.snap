// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runStorageMigrations does not run migrations when on the latest version 1.0.0 1`] = `
{
  "domainList": [],
  "exportLinks": [],
  "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
    "active": true,
    "identifier": "b7fce47e-58e8-4409-adf4-08da053e713d",
    "queries": [
      "my jaw dropped",
      "I was shocked",
    ],
    "queryPatterns": [],
    "replacement": "I was surprised",
  },
  "preferences": {
    "activeTab": "rules",
    "domainListEffect": "deny",
    "extensionEnabled": true,
    "focusRule": "",
  },
  "storageVersion": "1.0.0",
}
`;

exports[`runStorageMigrations does not run migrations when on the latest version 1.1.0 1`] = `
{
  "domainList": [],
  "exportLinks": [],
  "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
    "active": true,
    "identifier": "b7fce47e-58e8-4409-adf4-08da053e713d",
    "queries": [
      "my jaw dropped",
      "I was shocked",
    ],
    "queryPatterns": [],
    "replacement": "I was surprised",
    "useGlobalReplacementStyle": true,
  },
  "preferences": {
    "activeTab": "rules",
    "domainListEffect": "deny",
    "extensionEnabled": true,
    "focusRule": "",
  },
  "replacementStyle": {
    "active": false,
    "backgroundColor": "#ffc107",
    "color": "#dc3545",
    "options": [
      "backgroundColor",
    ],
  },
  "storageVersion": "1.1.0",
}
`;

exports[`runStorageMigrations migrates from 1.0.0 to 1.1.0 1`] = `
{
  "domainList": [],
  "exportLinks": [],
  "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
    "active": true,
    "identifier": "b7fce47e-58e8-4409-adf4-08da053e713d",
    "queries": [
      "my jaw dropped",
      "I was shocked",
    ],
    "queryPatterns": [],
    "replacement": "I was surprised",
    "useGlobalReplacementStyle": true,
  },
  "preferences": {
    "activeTab": "rules",
    "domainListEffect": "deny",
    "extensionEnabled": true,
    "focusRule": "",
  },
  "replacementStyle": {
    "active": false,
    "backgroundColor": "#ffc107",
    "color": "#dc3545",
    "options": [
      "backgroundColor",
    ],
  },
  "storageVersion": "1.1.0",
}
`;

exports[`runStorageMigrations migrates from undefined version to 1.0.0 1`] = `
{
  "domainList": [],
  "exportLinks": [],
  "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
    "active": true,
    "identifier": "b7fce47e-58e8-4409-adf4-08da053e713d",
    "queries": [
      "my jaw dropped",
      "I was shocked",
    ],
    "queryPatterns": [],
    "replacement": "I was surprised",
  },
  "preferences": {
    "activeTab": "rules",
    "domainListEffect": "deny",
    "extensionEnabled": true,
    "focusRule": "",
  },
  "storageVersion": "1.0.0",
}
`;

exports[`runStorageMigrations migrates from undefined version to 1.1.0 1`] = `
{
  "domainList": [],
  "exportLinks": [],
  "matcher__b7fce47e-58e8-4409-adf4-08da053e713d": {
    "active": true,
    "identifier": "b7fce47e-58e8-4409-adf4-08da053e713d",
    "queries": [
      "my jaw dropped",
      "I was shocked",
    ],
    "queryPatterns": [],
    "replacement": "I was surprised",
    "useGlobalReplacementStyle": true,
  },
  "preferences": {
    "activeTab": "rules",
    "domainListEffect": "deny",
    "extensionEnabled": true,
    "focusRule": "",
  },
  "replacementStyle": {
    "active": false,
    "backgroundColor": "#ffc107",
    "color": "#dc3545",
    "options": [
      "backgroundColor",
    ],
  },
  "storageVersion": "1.1.0",
}
`;
